### PR TITLE
fix: avoid creating too many externrefs

### PR DESCRIPTION
**Pull request by zshipko**
Updates plugins to allocate a single `ExternRef` for the host context up front, to avoid running into the `failed to allocate externref` error from Wasmtime

### CODE REVIEW

**Review by chrisdickinson — APPROVED** [review]
Oof, that's a tricky one! I left one comment to address – zeroing out `self.host_context` when we don't receive a host context – but otherwise I think this LGTM!

File: runtime/src/plugin.rs [hunk 1]
```diff
@@ -744,7 +758,22 @@ impl Plugin {
...
+                .host_context
+                .data_mut(&mut self.store)
+                .map_err(|x| (x, -1))?;
+            if let Some(inner) = inner.downcast_mut::<Box<dyn std::any::Any + Send + Sync>>() {
+                let x: Box<T> = Box::new(host_context);
+                *inner = x;
+            }
+            Some(self.host_context.clone())
+        } else {
+            None
```
[the remark this answers is not in the document]
Comment (zshipko) [reply runtime/src/plugin.rs:774]: I made it so we're zeroing out the host context after each call: https://github.com/extism/extism/pull/759/files/be744b5ea7df5c1a6f3d5accb28324c811a71e73#diff-fc3b26932af6a72d6a908c3d72663d22045526b2c46c6777aadeb7852d1137d4R816 - but am open to moving it up to here if you think that makes more sense.

File: runtime/src/tests/runtime.rs [hunk 2]
```diff
@@ -366,7 +368,7 @@ fn test_call_with_host_context() {
         [PTR],
         UserData::default(),
         |current_plugin, _val, ret, _user_data: UserData<()>| {
-            let foo = current_plugin.host_context::<Foo>()?;
+            let foo = current_plugin.host_context::<Foo>()?.clone();
             let hnd = current_plugin.memory_new(foo.message)?;
             ret[0] = current_plugin.memory_to_val(hnd);
             Ok(())
```
[the remark this answers is not in the document]
Comment (zshipko) [reply runtime/src/tests/runtime.rs:374]: The host context should be reset before the next call (see the comment below for the link)

File: runtime/src/plugin.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrisdickinson) [reply runtime/src/plugin.rs:774]: Oh, perfect! That makes sense.

**Pull request merged**